app/ante/cosmos/fees.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	if !simulate && ctx.BlockHeight() > 0 && feeTx.GetGas() == 0 {
+		return ctx, errorsmod.Wrap(errortypes.ErrInvalidGasLimit, "must provide positive gas")
+	}
+
+	var (
+		priority int64
+		err      error
+	)
+
+	fee := feeTx.GetFee()
+	if !simulate {
+		fee, priority, err = dfd.txFeeChecker(ctx, tx)
+		if err != nil {
+			return ctx, err
+		}
+	}
+	if err := dfd.checkDeductFee(ctx, tx, fee); err != nil {
+		return ctx, err
+	}
+
+	newCtx := ctx.WithPriority(priority)
+
+	return next(newCtx, tx, simulate)
+}
+
+// checkDeductFee checks if the fee payer has enough funds to pay for the fees and deducts them.
+// If the spendable balance is not enough, it tries to claim enough staking rewards to cover the fees.
+func (dfd DeductFeeDecorator) checkDeductFee(ctx sdk.Context, sdkTx sdk.Tx, fees sdk.Coins) error {
+	feeTx, ok := sdkTx.(sdk.FeeTx)
+	if !ok {
+		return errorsmod.Wrap(errortypes.ErrTxDecode, "Tx must implement the FeeTx interface")
+	}
+
+	if addr := dfd.accountKeeper.GetModuleAddress(authtypes.FeeCollectorName); addr == nil {
+		return fmt.Errorf("fee collector module account (%s) has not been set", authtypes.FeeCollectorName)
+	}
+
+	feePayer := feeTx.FeePayer()
+	feeGranter := feeTx.FeeGranter()
+	deductFeesFrom := feePayer
+
+	// if feegranter is set, then deduct the fee from the feegranter account.
+	// this works only when feegrant is enabled.
+	if feeGranter != nil {
+		if dfd.feegrantKeeper == nil {
+			return errortypes.ErrInvalidRequest.Wrap("fee grants are not enabled")
+		} else if !feeGranter.Equals(feePayer) {
+			err := dfd.feegrantKeeper.UseGrantedFees(ctx, feeGranter, feePayer, fees, sdkTx.GetMsgs())
+			if err != nil {
+				return errorsmod.Wrapf(err, "%s does not not allow to pay fees for %s", feeGranter, feePayer)
+			}
+		}
+
+		deductFeesFrom = feeGranter
+	}
+
+	deductFeesFromAcc := dfd.accountKeeper.GetAccount(ctx, deductFeesFrom)
+	if deductFeesFromAcc == nil {
+		return errortypes.ErrUnknownAddress.Wrapf("fee payer address: %s does not exist", deductFeesFrom)
+	}
+
+	// deduct the fees
+	if !fees.IsZero() {
+		if err := dfd.deductFeesFromBalanceOrUnclaimedRewards(ctx, deductFeesFromAcc, fees); err != nil {
+			return err
+		}
+	}
+
+	events := sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeTx,
+			sdk.NewAttribute(sdk.AttributeKeyFee, fees.String()),
+			sdk.NewAttribute(sdk.AttributeKeyFeePayer, deductFeesFrom.String()),
+		),
+	}
+	ctx.EventManager().EmitEvents(events)
+
+	return nil
+}
+
+// deductFeesFromBalanceOrUnclaimedRewards tries to deduct the fees from the account balance.
+// If the account balance is not enough, it tries to claim enough staking rewards to cover the fees.
+// If the account does not have sufficient staking rewards, it returns an error.
+func (dfd DeductFeeDecorator) deductFeesFromBalanceOrUnclaimedRewards(ctx sdk.Context, deductFeesFromAcc authtypes.AccountI, fees sdk.Coins) error {
+	balance := dfd.bankKeeper.GetBalance(ctx, deductFeesFromAcc.GetAddress(), fees[0].Denom)
```

Comment:
this could panic if someone provides more than one coin for the fees. You need to get the bond denom from the staking keeper and check for that amount